File: src/my_project/interviews_typescript/top_150_questions_round_1/ex_62_remove_duplicates_from_sorted_linked_list_ii.ts
import { ListNode } from './ListNode';   

function deleteDuplicates(head: ListNode | null): ListNode | null {
    // Create a dummy node to handle edge cases where head is removed
    const dummy = new ListNode(0);
    dummy.next = head;
    let prev = dummy;
    
    while (head) {
        // Check if current node has duplicates
        if (head.next && head.val === head.next.val) {
            // Skip all nodes with the same value
            while (head.next && head.val === head.next.val) {
                head = head.next;
            }
            // Link prev to the node after duplicates
            prev.next = head.next;
        } else {
            // No duplicate, move prev forward
            prev = prev.next!;
        }
        
        // Move to next node
        head = head.next;
    }
    
    return dummy.next;
}

File: src/my_project/interviews/top_150_questions_round_22/ex_62_remove_duplicates_from_sorted_linked_list_ii.py
from typing import List, Union, Collection, Mapping, Optional
from abc import ABC, abstractmethod

# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def deleteDuplicates(self, head: Optional[ListNode]) -> Optional[ListNode]:
        # Create a dummy node to handle edge cases where head is removed
        dummy = ListNode(0)
        dummy.next = head
        prev = dummy
        
        while head:
            # Check if current node has duplicates
            if head.next and head.val == head.next.val:
                # Skip all nodes with the same value
                while head.next and head.val == head.next.val:
                    head = head.next
                # Link prev to the node after duplicates
                prev.next = head.next
            else:
                # No duplicate, move prev forward
                prev = prev.next
            
            # Move to next node
            head = head.next
        
        return dummy.next